xds/src/main/java/io/grpc/xds/AbstractXdsClient.java

```diff
@@ -0,0 +1,760 @@
+  }
+
+  /**
+   * Called when the ADS stream is closed passively.
+   */
+  // Must be synchronized.
+  protected void handleStreamClosed(Status error) {
+  }
+
+  /**
+   * Called when the ADS stream has been recreated.
+   */
+  // Must be synchronized.
+  protected void handleStreamRestarted() {
+  }
+
+  /**
+   * Called when being shut down.
+   */
+  // Must be synchronized.
+  protected void handleShutdown() {
+  }
+
+  /**
+   * Synchronizes the execution of the given {@code runnable} with other state mutating operations.
+   */
+  protected abstract void runWithSynchronized(Runnable runnable);
+
+  @Override
+  final void shutdown() {
+    runWithSynchronized(new Runnable() {
+      @Override
+      public void run() {
+        shutdown = true;
+        logger.log(XdsLogLevel.INFO, "Shutting down");
+        xdsChannel.getManagedChannel().shutdown();
+        if (adsStream != null) {
+          adsStream.close(Status.CANCELLED.withDescription("shutdown").asException());
+        }
+        if (rpcRetryTimer != null) {
+          rpcRetryTimer.cancel(false);
+        }
+        handleShutdown();
+      }
+    });
+  }
+
+  @Override
+  public String toString() {
+    return logId.toString();
+  }
+
+  /**
+   * Returns the collection of resources currently subscribing to or {@code null} if not
+   * subscribing to any resources for the given type.
+   *
+   * <p>Note a empty collection indicates subscribing to resources of the given type with
+   * wildcard mode.
+   */
+  // Must be synchronized.
+  @Nullable
+  abstract Collection<String> getSubscribedResources(ResourceType type);
+
+  /**
+   * Updates the resource subscription for the given resource type.
+   */
+  // Must be synchronized.
+  protected final void adjustResourceSubscription(ResourceType type) {
+    if (isInBackoff()) {
+      return;
+    }
+    if (adsStream == null) {
+      startRpcStream();
+    }
+    Collection<String> resources = getSubscribedResources(type);
+    if (resources != null) {
+      adsStream.sendDiscoveryRequest(type, resources);
+    }
+  }
+
+  /**
+   * Accepts the update for the given resource type by updating the latest resource version
+   * and sends an ACK request to the management server.
+   */
+  // Must be synchronized.
+  protected final void ackResponse(ResourceType type, String versionInfo, String nonce) {
+    switch (type) {
+      case LDS:
+        ldsVersion = versionInfo;
+        break;
+      case RDS:
+        rdsVersion = versionInfo;
+        break;
+      case CDS:
+        cdsVersion = versionInfo;
+        break;
+      case EDS:
+        edsVersion = versionInfo;
+        break;
+      case UNKNOWN:
+      default:
+        throw new AssertionError("Unknown resource type: " + type);
+    }
+    logger.log(XdsLogLevel.INFO, "Sending ACK for {0} update, nonce: {1}, current version: {2}",
+        type, nonce, versionInfo);
+    Collection<String> resources = getSubscribedResources(type);
+    if (resources == null) {
```

Comment:
Behavior of `getSubscribedResources` as I have understood is:

- the one in `ServerXdsClient` returns null if no watcher set else emptyList
- the one in `ClientXdsClient` returns null instead of emptyList
- over here if it is `null` you change it back to emptyList

Why can't we then have `getSubscribedResources` return emptyList instead of null? So null should never be returned unless I am missing something

Reply:
The problem really comes from the `ServerXdsClient`'s implementation. As you mentioned, returning `null` and empty list means different things for `ServerXdsClient`: `null` indicates nothing is being subscribed to and when the ADS stream is recreated, no request should be sent. While empty list does mean it is currently subscribing to the wildcard Listener resource, and a request need to be sent when the ADS stream is recreated.

`ClientXdsClient`'s implementation will always return `null` if nothing being subscribed to, or a non-empty list. It has to be consistent with `ServerXdsClient`'s semantics. It cannot return an empty list as its semantics has alreayd been taken. The abstract layer does not know if the real implementation is a `ServerXdsClient` or a `ClientXdsClient`.

Reply:
Okay, thanks for the clarification. I thought the client side (ClientXdsClient) also uses wildcard queries (since they are supported for clusters and listeners) but it looks like our client-side never needs to issue wild-card queries because of the specific use-case of proxyless gRPC (unlike Envoy which uses wildcard queries to get all the clusters and listeners in the mesh).

The server side will also now use specific resources (assuming TD implements the new API) so the emptyList case should never happen and this code can then be cleaned up.